delta-app/src/main/java/io/cdap/delta/app/DeltaContext.java

```diff
@@ -39,23 +44,34 @@
  */
 public class DeltaContext implements DeltaSourceContext, DeltaTargetContext {
   private static final String STATE_PREFIX = "state-";
+  private static final Logger LOG = LoggerFactory.getLogger(DeltaContext.class);
   private final DeltaPipelineId id;
   private final String runId;
   private final Metrics metrics;
   private final StateStore stateStore;
   private final PluginContext pluginContext;
   private final EventMetrics eventMetrics;
   private final PipelineStateService stateService;
+  private final RetryPolicy<Object> commitOffsetRetryPolicy;
 
-  public DeltaContext(DeltaPipelineId id, String runId, Metrics metrics, StateStore stateStore,
-                      PluginContext pluginContext, EventMetrics eventMetrics, PipelineStateService stateService) {
+  DeltaContext(DeltaPipelineId id, String runId, Metrics metrics, StateStore stateStore,
+               PluginContext pluginContext, EventMetrics eventMetrics, PipelineStateService stateService) {
     this.id = id;
     this.runId = runId;
     this.metrics = metrics;
     this.stateStore = stateStore;
     this.pluginContext = pluginContext;
     this.eventMetrics = eventMetrics;
     this.stateService = stateService;
+    this.commitOffsetRetryPolicy = new RetryPolicy<>()
+      .withBackoff(1, 120, ChronoUnit.SECONDS)
+      .withMaxAttempts(Integer.MAX_VALUE)
```

Comment:
Should we have an upper bound for the max num of backoff retry? The current max num of attempts are so high, what if something bad happened, the app will stuck for very long time. Looks like it is better to put some reasonable upper bound, so we can fail the pipeline quicker.

Reply:
yeah I changed my mind a few times about what the right thing to do here would be. The problem is that there isn't a good way to fail the pipeline when this fails, because commitOffset() is called by the target plugin. So the app doesn't have any way to fail the pipeline here.

Now I'm thinking that it might be better to remove the retry and let the plugin handle it however it wants.

Reply:
Going to remove the retry here and force the plugin to handle it. The BQ target isn't going to want to throw the IOException from applyDDL or applyDML, as that would cause offset to be reset, which would cause a batch of changes to be replayed before failing again on commit and repeating the process forever.

To handle this, I'm going to add a special type of Exception that plugins can throw that will tell the app not to retry and fail immediately. This will allow the target plugin to retry the commit themselves for some time before throwing that special exception and failing the pipeline.

Reply:
Sounds good.